oxidation/libparsec/crates/client_types/src/logged_core.rs

```diff
@@ -0,0 +1,53 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use once_cell::sync::OnceCell;
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+    sync::Mutex,
+};
+
+use libparsec_types::DeviceID;
+
+static CORE: OnceCell<Mutex<HashMap<i32, LoggedCore>>> = OnceCell::new();
+
+#[derive(Debug, Clone)]
+struct LoggedCore {
+    // TODO: Remove me when LoggedCore is implemented
+    test_device_id: DeviceID,
+}
+
+// We don't need to hash everything !
+impl Hash for LoggedCore {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.test_device_id.hash(state)
+    }
+}
+
+/// 32 bits of signature seems to be enough, people don't have that much devices
+/// Output: i32 is because javascript number is a 64 bits float
+pub fn login(test_device_id: DeviceID) -> i32 {
+    let core = LoggedCore { test_device_id };
+
+    // Make a signature for LoggedCore
+    let mut hasher = DefaultHasher::new();
+    core.hash(&mut hasher);
+    let key = hasher.finish() as i32;
```

Comment:
There is no guarantee the hash is unique, and the fact we cut the hash to 32bits increase this risk.

Instead of doing a hash, use a global vector of Option<LoggedCore>.
On logging you iterate over the vector to check the device id is not already logged, then push the new LoggedCore.
This way your id is the index in the vector which is super convenient.
For log out you just replace the element in the vector by None.

This means the vector only keep growing (so you never reuse an id which is good to avoid hard to track bugs), and the impact is totally negligible on memory usage (one new element in the vector per login)

Reply:
Done